File: app/src/main/java/com/readrops/app/database/dao/FolderDao.java
package com.readrops.app.database.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import com.readrops.app.database.entities.Account;
import com.readrops.app.database.entities.Folder;

import java.util.ArrayList;
import java.util.List;

@Dao
public abstract class FolderDao implements BaseDao<Folder> {

    @Query("Select * from Folder Where account_id = :accountId Order By name ASC")
    public abstract LiveData<List<Folder>> getAllFolders(int accountId);

    @Query("Select * from Folder Where account_id = :accountId Order By name ASC")
    public abstract List<Folder> getFolders(int accountId);

    @Query("Update Folder set name = :name Where remoteId = :remoteFolderId And account_id = :accountId")
    public abstract void updateName(String remoteFolderId, int accountId, String name);

    @Query("Select case When :remoteId In (Select remoteId From Folder Where account_id = :accountId) Then 1 else 0 end")
    abstract boolean remoteFolderExists(String remoteId, int accountId);

    @Query("Select * from Folder Where id = :folderId")
    public abstract Folder select(int folderId);

    @Query("Select remoteId From Folder Where account_id = :accountId")
    public abstract List<String> getFolderRemoteIdsOfAccount(int accountId);

    @Query("Delete From Folder Where id in (:ids)")
    abstract void deleteByIds(List<String> ids);

    /**
     * Insert, update and delete folders
     *
     * @param folders folders to insert or update
     * @param account owner of the feeds
     * @return the list of the inserted feeds ids
     */
    @Transaction
    public List<Long> foldersUpsert(List<Folder> folders, Account account) {
        List<String> accountFolderIds = getFolderRemoteIdsOfAccount(account.getId());
        List<Folder> foldersToInsert = new ArrayList<>();

        for (Folder folder : folders) {
            if (remoteFolderExists(folder.getRemoteId(), account.getId())) {
                updateName(folder.getRemoteId(), account.getId(), folder.getName());

                accountFolderIds.remove(folder.getRemoteId());
            } else {
                foldersToInsert.add(folder);
            }
        }

        if (!accountFolderIds.isEmpty())
            deleteByIds(accountFolderIds);

        return insert(foldersToInsert);
    }
}

File: db/src/main/java/com/readrops/db/dao/FolderDao.kt
package com.readrops.db.dao

import androidx.room.Dao
import androidx.room.Query
import androidx.room.RawQuery
import androidx.room.Transaction
import androidx.sqlite.db.SupportSQLiteQuery
import com.readrops.db.entities.Feed
import com.readrops.db.entities.Folder
import com.readrops.db.entities.Item
import com.readrops.db.entities.account.Account
import com.readrops.db.pojo.FolderWithFeed
import kotlinx.coroutines.flow.Flow

@Dao
interface FolderDao : BaseDao<Folder> {

    // TODO react to Item changes when this table is not part of the query might be a perf issue
    @RawQuery(observedEntities = [Folder::class, Feed::class, Item::class])
    fun selectFoldersAndFeeds(query: SupportSQLiteQuery): Flow<List<FolderWithFeed>>

    @Query("Select * From Folder Where account_id = :accountId")
    fun selectFolders(accountId: Int): Flow<List<Folder>>

    @Query("Select * from Folder Where id = :folderId")
    fun select(folderId: Int): Folder

    @Query("Select * From Folder Where name = :name And account_id = :accountId")
    suspend fun selectFolderByName(name: String, accountId: Int): Folder?

    @Query("Select remoteId From Folder Where account_id = :accountId")
    suspend fun selectFolderRemoteIds(accountId: Int): List<String>

    @Query("Update Folder set name = :name Where remoteId = :remoteId And account_id = :accountId")
    suspend fun updateFolderName(name: String, remoteId: String, accountId: Int)

    @Query("Delete From Folder Where remoteId in (:ids) And account_id = :accountId")
    suspend fun deleteByIds(ids: List<String>, accountId: Int)

    /**
     * Insert, update and delete folders
     *
     * @param folders folders to insert or update
     * @param account owner of the feeds
     * @return the list of the inserted folders ids
     */
    @Transaction
    suspend fun upsertFolders(folders: List<Folder>, account: Account): List<Long> {
        val localFolderIds = selectFolderRemoteIds(account.id)

        val foldersToInsert = folders.filter { folder -> localFolderIds.none { localFolderId -> folder.remoteId == localFolderId  } }
        val foldersToDelete = localFolderIds.filter { localFolderId -> folders.none { folder -> localFolderId == folder.remoteId } }

        // folders to update
        folders.filter { folder -> localFolderIds.any { localFolderId -> folder.remoteId == localFolderId} }
            .forEach { updateFolderName(it.name!!, it.remoteId!!, account.id) }

        if (foldersToDelete.isNotEmpty()) {
            deleteByIds(foldersToDelete, account.id)
        }

        return insert(foldersToInsert)
    }
}